### DISCUSSION TITLE
density_grid_bitfield and density_grid_mean

### DISCUSSION THREAD

**Discussion by neilgogogo** [NVlabs/instant-ngp#353] [General]
Hi, I read the code and I got lost in the density_grid_bitfield and density grid mean.

I am confused about what is density_grid_bitfield. As the size is density_grid.size()/8. It seems to combines 8 grids into one big grid and to test whether the 8 smaller grids in the bigger cube are larger than the threshold. 

I am also quite confused about the density_grid_mean which is used for the threshold after min(0.01f, density_grid_mean). It is calculated after the sum_reduce operation with the input of density_grid. But I am a little confused about how it is calculated.

https://github.com/NVlabs/instant-ngp/blob/833f7303c3fadb81927c765cd67788209cf66f1c/src/testbed_nerf.cu#L2445
<img width="1201" alt="image" src="https://user-images.githubusercontent.com/53148705/158589188-d30b8862-9d5f-4693-b6e1-71c3a3c4cd94.png">

Many thanks.

**Comment by Tom94** [comment]
Hi there,

the occupancy bitfield uses a single bit per cell of the density grid to either mark that cell as occupied (1) or as empty (0). As such, one byte (8 bits) can hold the occupancy of 8 cells of the density grid, hence the division by 8.

The mean computation is your typical parallel sum reduction (along with a division by the number of elements to turn the sum into a mean). Parallel reduction on the GPU is somewhat tricky to make efficient, which is probably why the implementation looks confusing. It's ultimately a mish-mash of some of the techniques described here: https://developer.download.nvidia.com/assets/cuda/files/reduction.pdf

Lastly, `bitfield_max_pool` indeed combines blocks of 8 values into a single, lower-resolution value. It's used to make the coarser cascades more accurate in regions where finer cascades are also present.

Cheers!
